components/autocomplete/index.js

```diff
@@ -225,22 +225,85 @@ export class Autocomplete extends Component {
 		this.node = node;
 	}
 
-	insertCompletion( range, replacement ) {
-		const container = document.createElement( 'div' );
-		container.innerHTML = renderToString( replacement );
-		while ( container.firstChild ) {
-			const child = container.firstChild;
-			container.removeChild( child );
-			range.insertNode( child );
-			range.setStartAfter( child );
+	insertCompletion( range, replacement, completerName ) {
+		// Wrap completions so we can treat them as tokens with editing boundaries.
+		const tokenWrapper = document.createElement( 'span' );
+		tokenWrapper.innerHTML = renderToString( replacement );
+
+		// Remember the completer in case the user wants to edit the token.
+		tokenWrapper.dataset.autocompleter = completerName;
+
+		// Add classes for general and completer-specific styling.
+		tokenWrapper.classList.add( 'autocomplete-token' );
+		tokenWrapper.classList.add( `autocomplete-token-${ completerName }` );
+
+		const existingTokenWrapper = this.getTokenWrapperNode( range.startContainer );
+		if ( existingTokenWrapper ) {
+			/**
+			 * If we're within an existing token, we want to replace it rather
+			 * than inserting a completion within a completion.
+			 */
+			existingTokenWrapper.parentNode.replaceChild( tokenWrapper, existingTokenWrapper );
+		} else {
+			range.insertNode( tokenWrapper );
 		}
+
+		range.setStartAfter( tokenWrapper );
 		range.deleteContents();
+
+		const selection = window.getSelection();
+		selection.removeAllRanges();
+
+		const newCursorPosition = document.createRange();
+		/**
+		 * Add a zero-width non-breaking space (ZWNBSP) so we can place cursor
+		 * after. TinyMCE handles ZWNBSP's nicely around token boundaries,
+		 * adding and removing them as necessary as the keyboard is used to
+		 * move the cursor in and out of boundaries.
+		 */
+		tokenWrapper.parentNode.insertBefore(
+			document.createTextNode( '\uFEFF' ),
+			tokenWrapper.nextSibling
+		);
+		newCursorPosition.setStartAfter( tokenWrapper.nextSibling );
+		selection.addRange( newCursorPosition );
+	}
+
+	/**
+	 * Gets the token wrapper for a node if it has one.
+	 *
+	 * @param {Node} possibleTokenChild The node for which to find the wrapper.
+	 *
+	 * @return {Node?} The token wrapper or null if there is none.
+	 * @private
+	 */
+	getTokenWrapperNode( possibleTokenChild ) {
+		const element = isTextNode( possibleTokenChild ) ?
+			possibleTokenChild.parentNode :
+			possibleTokenChild;
+		const withinToken = element.matches( '.autocomplete-token, .autocomplete-token *' );
```

Comment:
Is the dependency defined in this file? See also #7159.

Reply:
Thanks for the reminder. I missed that.

I added import of 'element-closest' to add the `Element#matches` polyfill for now, but I'm hoping to work on a better way of registering and including polyfills later this week.